src/backend/parser/parse_agg.c

```diff
@@ -842,33 +847,45 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
 	/*
 	 * Build a list of the acceptable GROUP BY expressions for use by
 	 * check_ungrouped_columns().
+	 *
+	 * We get the TLE, not just the expr, because GROUPING wants to know the
+	 * sortgroupref.
 	 */
 	foreach(l, qry->groupClause)
 	{
 		Node	   *grpcl = lfirst(l);
-		List	   *exprs;
+		List	   *tles;
 		ListCell   *l2;
 
 		if (grpcl == NULL)
 			continue;
 
 		Assert(IsA(grpcl, SortGroupClause) || IsA(grpcl, GroupingClause));
 
-		exprs = get_groupclause_exprs(grpcl, qry->targetList);
+		tles = get_groupclause_tles(grpcl, qry->targetList);
 
-		foreach(l2, exprs)
+		foreach(l2, tles)
 		{
-			Node	   *expr = (Node *) lfirst(l2);
+			TargetEntry *tl = (TargetEntry*) lfirst(l2);
 
 			// FIXME: Should this go into check_agg_arguments now?
-			if (checkExprHasGroupExtFuncs(expr))
+			if (checkExprHasGroupExtFuncs((Node*)tl->expr))
 				ereport(ERROR,
 						(errcode(ERRCODE_GROUPING_ERROR),
 						 errmsg("grouping() or group_id() not allowed in GROUP BY clause")));
-			groupClauses = lcons(expr, groupClauses);
+			groupClauses = lcons(tl, groupClauses);
 		}
 	}
 
+	/*
+	 * Getting adjacent TargetEntrys in groupings. Common tle will be needed
```

Comment:
I doubt that the 'adjacent' is a good term.
And what is a 'common ressortgroupref' should be explicitly explained in the comment.

Reply:
Done. I hope the new description is exhaustive.